### PR TITLE
ESP32: Rotating device ID for BLE transport

### LINKED ISSUE #5370: No provision for adding GATT-based Additional Data
#### Problem
As per the CHIP specification section 5.2.3.8.6 Advertising Data, the table shows that for byte 14, the bit [0] should be additional data flag and bits [7:1] are reserved and set to 0.  In the CHIP code, I could not find that being added. As per the spec, the length of AD Type 0x16 is 11 (0x0b), but due to the above byte being missing, it is seen as 10 (0x0a).

 #### Proposed Solution
Need to add provision to support GATT-based Additional Data.

### PR DESCRIPTION
**Pull request by shubhamdp**
#### Problem
- Rotating device ID is not supported on ESP32 platform
- Fixes #5370

#### Change overview
- Added additional data flag in BLE service data
- Added support for rotational device id for ESP32 platform
- This is an optional feature which can be controlled using config opitons 'chip_enable_additional_data_advertising' and 'chip_enable_rotating_device_id'

#### Testing
- Tested manually on Linux and ESP32 platforms using nrfconnect android app
- On enabling the config options, advertising packet shows an additional byte for additional data flag.
- It also shows an extra C3 characteristic. When read the characteristic it shows the additional data containing rotational device id.

### CODE REVIEW

**Review by gjc13 — APPROVED** [review]

File: src/ble/CHIPBleServiceData.h [hunk 1]
```diff
@@ -46,11 +46,17 @@ enum chipBLEServiceDataType
...
     constexpr static uint16_t kDiscriminatorMask = 0xfff;
+#if CHIP_ENABLE_ADDITIONAL_DATA_ADVERTISING
+    constexpr static uint8_t kAdditionalDataFlagMask = 0x1;
+#endif
 
     uint8_t OpCode;
     uint8_t DeviceDiscriminator[2];
     uint8_t DeviceVendorId[2];
     uint8_t DeviceProductId[2];
+#if CHIP_ENABLE_ADDITIONAL_DATA_ADVERTISING
```
Comment (Damian-Nordic) [inline src/ble/CHIPBleServiceData.h:57]: This should be always present in the advertising data. It should only be set differently depending on whether the C3 characteristic is present or not. Therefore, the ifdefs can be removed in this header.

Comment (shubhamdp) [reply src/ble/CHIPBleServiceData.h]: @Damian-Nordic, I removed the ifdef. Now, for Linux and ESP32, if config option "CHIP_ENABLE_ADDITIONAL_DATA_ADVERTISING" is enabled, we set the Additional Data Flag and start showing C3 characteristic.

I also increased bumped the static assert check for size of BLE advertisement data in src/platform/Zephyr/BLEManagerImpl.cpp, PTAL!

File: src/app/server/Dnssd.cpp [hunk 2]
```diff
@@ -338,7 +338,7 @@ CHIP_ERROR DnssdServer::Advertise(bool commissionableNode, chip::Dnssd::Commissi
 #if CHIP_ENABLE_ROTATING_DEVICE_ID
     char rotatingDeviceIdHexBuffer[RotatingDeviceId::kHexMaxLength];
     ReturnErrorOnFailure(GenerateRotatingDeviceId(rotatingDeviceIdHexBuffer, ArraySize(rotatingDeviceIdHexBuffer)));
-    advertiseParameters.SetRotatingId(chip::Optional<const char *>::Value(rotatingDeviceIdHexBuffer));
+    advertiseParameters.SetRotatingDeviceId(chip::Optional<const char *>::Value(rotatingDeviceIdHexBuffer));
```
Comment (dhrishi) [inline src/app/server/Dnssd.cpp:341]: This change is not required now. It is now merged in https://github.com/project-chip/connectedhomeip/pull/13123

File: src/platform/BUILD.gn [hunk 3]
```diff
@@ -54,7 +54,7 @@ if (chip_device_platform != "none") {
     chip_bypass_rendezvous = false
 
     # Enable including the additional data in the advertisement packets
-    chip_enable_additional_data_advertising = true
+    chip_enable_additional_data_advertising = false
```
Comment (dhrishi) [inline src/platform/BUILD.gn:57]: Please confirm that it does not affect any other platforms

File: src/platform/ESP32/BUILD.gn [hunk 4]
```diff
@@ -48,7 +48,10 @@ static_library("ESP32") {
     "nimble/BLEManagerImpl.cpp",
   ]
 
-  deps = [ "${chip_root}/src/lib/dnssd:platform_header" ]
+  deps = [
+    "${chip_root}/src/lib/dnssd:platform_header",
+    "${chip_root}/src/setup_payload",
```
Comment (dhrishi) [inline src/platform/ESP32/BUILD.gn:53]: Why is this required?

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 5]
```diff
@@ -90,6 +91,17 @@ const ble_uuid128_t UUID_CHIPoBLEChar_TX   = {
...
 };
 
+#if CHIP_ENABLE_ADDITIONAL_DATA_ADVERTISING
+const ble_uuid128_t UUID128_CHIPoBLEChar_C3 = {
+    BLE_UUID_TYPE_128, { 0x04, 0x8F, 0x21, 0x83, 0x8A, 0x74, 0x7D, 0xB8, 0xF2, 0x45, 0x72, 0x87, 0x38, 0x02, 0x63, 0x64 }
+};
+const ChipBleUUID chipUUID_CHIPoBLEChar_C3 = { { 0x64, 0x63, 0x02, 0x38, 0x87, 0x72, 0x45, 0xF2, 0xB8, 0x7D, 0x74, 0x8A, 0x83, 0x21,
+                                                 0x8F, 0x04 } };
+const ble_uuid128_t UUID_CHIPoBLEChar_C3   = {
+    { BLE_UUID_TYPE_128 }, { 0x04, 0x8F, 0x21, 0x83, 0x8A, 0x74, 0x7D, 0xB8, 0xF2, 0x45, 0x72, 0x87, 0x38, 0x02, 0x63, 0x64 }
```
Comment (dhrishi) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:101]: Isn't it the same as `UUID128_CHIPoBLEChar_C3` above?

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 6]
```diff
@@ -683,6 +706,11 @@ CHIP_ERROR BLEManagerImpl::ConfigureAdvertisingData(void)
     uint8_t advData[MAX_ADV_DATA_LEN];
     uint8_t index = 0;
 
+    constexpr uint8_t kServiceDataLenSize  = 1;
```
Comment (dhrishi) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:709]: Should this be `kServiceDataTypeSize` instead?

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 5 above)
Comment (shubhamdp) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp:87]: Removed this as well as unused characteristic definitions.

File: src/platform/BUILD.gn (hunk 3 above)
Comment (shubhamdp) [reply src/platform/BUILD.gn:57]: Added `chip_enable_additional_data_advertising = true` to tv-app and tv-cast-app.

**Review by dhrishi — APPROVED** [review]

**Comment by dhrishi** [comment]
@chrisdecenzo @hnnajh PTAL

**Review by LuDuda — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
